src/librustc_codegen_llvm/intrinsic.rs

```diff
@@ -1039,6 +1039,29 @@ fn generic_simd_intrinsic(
     llret_ty: &'ll Type,
     span: Span
 ) -> Result<&'ll Value, ()> {
+    // Given a SIMD vector type `x` return the element type and the number of
+    // elements in the vector.
+    fn simd_type_size(bx: &Builder<'a, 'll, 'tcx>, x: Ty<'tcx>) -> (Ty<'tcx>, usize) {
+        let ty = if let ty::Adt(def, substs) = x.sty {
+            assert!(def.repr.simd());
+            let field0_ty = def.non_enum_variant().fields[0].ty(bx.tcx(), substs);
+            if let ty::Array(element_ty, _) = field0_ty.sty {
+                element_ty
+            } else {
+                field0_ty
+            }
+        } else {
+            bug!("should only be called with a SIMD type")
+        };
+        let count = if let ty::layout::Abi::Vector{count, ..} = bx.layout_of(x).abi {
```

Comment:
Might be a good idea to use the element `Scalar` from the `Abi::Vector` instead of the real type, but it would probably cause too much noise in this PR (if it can even work).

Reply:
Does Scalar give us the element type ? We need to at least distinguish between these types: `f{32,64}`, `{i,u}{8,16,32,64,128, size}`, `for<T> *mut T`, `for<T> *const T`.

Reply:
Really depends what it's used for. If it's for telling apart integer/float/pointer, then yeah, that's what `Scalar` does.

Reply:
I don't see how I can use scalar to, e.g. tell apart a `*mut *mut usize` from a `*const i32`. 

Maybe you could show an example ?

Reply:
No, you can't, but I don't know what the type is used for here.

Reply:
> No, you can't, but I don't know what the type is used for here.

This function generates LLVM-IR for all the generic simd intrinsics. The LLVM-IR for packed SIMD vectors includes the element type, e.g., `<4 x float**>`, so we need to be able to tell whether we have a `float*`, `float**`, `float***`, etc.

Reply:
We also use the `*const` vs `*mut` distinction, e.g., when generating monomorphization errors, e.g., we don't want to allow writes via `*const`.

Reply:
Look in `type_of.rs` for how that's done, I guess? I don't remember exactly, but we could offer something similar to `scalar_pair_element_llvm_type`, except for vectors.

I guess this is the kind of place where a signature like `ptrs: Simd<*mut T, N>, vals: Simd<T, N>` would be *really nice*, since you can rely on the typesystem for correctness.

Oh, bonus: `shuffle(a: Simd<T, A>, b: Simd<T, B>, indices: [usize; N]) -> Simd<T, N>` :D.

Reply:
We discussed this on Discord where I think the conclusion was that we do need to obtain the exact Rust field type to deal with some of the intrinsics, right?

Reply:
Yeah I think we just need to postpone this on having a `Simd<T, N>` type.

Reply:
So I've opened an issue to track doing this: https://github.com/rust-lang/rust/issues/63633